mteb/models/e5_instruct.py

```diff
@@ -0,0 +1,230 @@
+from transformers import AutoModel, AutoTokenizer, BatchEncoding
+from transformers.modeling_outputs import ModelOutput
+
+from mteb.encoder_interface import Encoder
+from mteb.model_meta import ModelMeta
+
+from .e5_models import E5_PAPER_RELEASE_DATE, XLMR_LANGUAGES
+from .instructions import task_to_instruction
+
+logger = logging.getLogger(__name__)
+
+T = TypeVar("T")
+EncodeTypes = Literal["query", "passage"]
+
+MISTRAL_LANGUAGES = ["eng_Latn", "fra_Latn", "deu_Latn", "ita_Latn", "spa_Latn"]
+
+
+def batched(iterable: Iterable[T], n: int) -> Iterable[tuple[T, ...]]:
+    """batched('ABCDEFG', 3) --> ABC DEF G"""  # noqa
+    if n < 1:
+        raise ValueError("n must be at least one")
+    it = iter(iterable)
+    while batch := tuple(islice(it, n)):
+        yield batch
+
+
+class E5InstructWrapper(Encoder):
+    def __init__(
+        self,
+        model_name: str,
+        revision: str,
+        max_length: int,
+        max_batch_size: Optional[int] = None,
+        **kwargs: Any,
+    ):
+        logger.info("Started loading e5 instruct model")
+        self.tokenizer = AutoTokenizer.from_pretrained(
+            model_name, revision=revision, **kwargs
+        )
+        self.model = AutoModel.from_pretrained(model_name, **kwargs)
+        self.max_length = max_length
+        self.max_batch_size = max_batch_size
+
+    def preprocess(
+        self, sentences: Sequence[str], instruction: str, encode_type: EncodeTypes
+    ) -> BatchEncoding:
+        if encode_type == "query":
+            sentences = [
+                f"Instruction: {instruction}\nQuery: {sentence}"
+                for sentence in sentences
+            ]
+
+        batch_dict = self.tokenizer(
+            sentences,  # type: ignore
+            max_length=512,
+            padding=True,
+            truncation=True,
+            return_tensors="pt",
+        )
+
+        return batch_dict.to(self.model.device)
+
+    def get_embedding_from_output(
+        self, output: ModelOutput, batch_dict: BatchEncoding
+    ) -> torch.Tensor:
+        return self.average_pool(output.last_hidden_state, batch_dict["attention_mask"])  # type: ignore
+
+    @staticmethod
+    def average_pool(
+        last_hidden_states: torch.Tensor, attention_mask: torch.Tensor
+    ) -> Tensor:
+        last_hidden = last_hidden_states.masked_fill(
+            ~attention_mask[..., None].bool(), 0.0
+        )
+        return last_hidden.sum(dim=1) / attention_mask.sum(dim=1)[..., None]
+
+    def encode(
+        self,
+        sentences: list[str],
+        *,
+        prompt_name: str | None = None,
+        batch_size: int = 32,
+        encode_type: EncodeTypes = "query",
+        **kwargs: Any,  # noqa
+    ) -> np.ndarray:
+        if self.max_batch_size and batch_size > self.max_batch_size:
+            batch_size = self.max_batch_size
+        batched_embeddings = []
+        if prompt_name is not None:
+            instruction = task_to_instruction(prompt_name)
```

Comment:
Nice, I think I'll also add this to GritLM after this PR though I'd have to subclass it I think and turn the `prompt_name` kwarg into the `instruction` kwarg used for GritLM (https://github.com/ContextualAI/gritlm/blob/da37ccbace1aa4f4bf26273e4e9a7cea705ae951/gritlm/gritlm.py#L98)